modules/sitemaps/sitemap-builder.php

```diff
@@ -0,0 +1,1843 @@
+<!-- generator='jetpack-{$jetpack_version}' -->
+<?xml-stylesheet type='text/xsl' href='{$sitemap_index_xsl_url}'?>
+<sitemapindex xmlns='http://www.sitemaps.org/schemas/sitemap/0.9'>
+HEADER
+			,
+			<<<FOOTER
+</sitemapindex>\n
+FOOTER
+			,
+			/* epoch */
+			'1970-01-01 00:00:00'
+		);
+
+		$page = $this->build_page_sitemap_tree();
+		$buffer->try_to_add_item( self::array_to_xml_string(
+			array(
+				'sitemap' => array(
+					'loc'     => $this->finder->construct_sitemap_url( $page['filename'] ),
+					'lastmod' => $page['last_modified'],
+				),
+			)
+		) );
+
+		$image = $this->build_image_sitemap_tree();
+		if ( false !== $image ) {
+			$buffer->try_to_add_item( self::array_to_xml_string(
+				array(
+					'sitemap' => array(
+						'loc'     => $this->finder->construct_sitemap_url( $image['filename'] ),
+						'lastmod' => $image['last_modified'],
+					),
+				)
+			) );
+		}
+
+		$video = $this->build_video_sitemap_tree();
+		if ( false !== $video ) {
+			$buffer->try_to_add_item( self::array_to_xml_string(
+				array(
+					'sitemap' => array(
+						'loc'     => $this->finder->construct_sitemap_url( $video['filename'] ),
+						'lastmod' => $video['last_modified'],
+					),
+				)
+			) );
+		}
+
+		$this->librarian->store_sitemap_data(
+			Jetpack_Sitemap_Librarian::MASTER_SITEMAP_NAME,
+			Jetpack_Sitemap_Librarian::MASTER_SITEMAP_TYPE,
+			$buffer->contents(),
+			''
+		);
+
+		if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+			$this->logger->time( 'End sitemap generation.' );
+		}
+
+		return;
+	}
+
+	/**
+	 * Build the page sitemap tree structure.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 *
+	 * @return array $args {
+	 *     @type string filename The filename of the root page sitemap.
+	 *     @type string last_modified The timestamp of the root page sitemap.
+	 * }
+	 */
+	private function build_page_sitemap_tree() {
+		$num_sitemaps = $this->build_all_page_sitemaps();
+
+		// If there's only one sitemap, make that the root.
+		if ( 1 === $num_sitemaps ) {
+			$this->librarian->delete_numbered_sitemap_rows_after(
+				Jetpack_Sitemap_Librarian::SITEMAP_INDEX_NAME_PREFIX,
+				0,
+				Jetpack_Sitemap_Librarian::SITEMAP_INDEX_TYPE
+			);
+
+			$last_modified = $this->librarian->read_sitemap_data(
+				Jetpack_Sitemap_Librarian::SITEMAP_NAME_PREFIX . '1',
+				Jetpack_Sitemap_Librarian::SITEMAP_TYPE
+			)['timestamp'];
+
+			return array(
+				'filename'      => Jetpack_Sitemap_Librarian::SITEMAP_NAME_PREFIX . '1.xml',
+				'last_modified' => str_replace( ' ', 'T', $last_modified ) . 'Z',
+			);
+		}
+
+		// Otherwise, we have to generate sitemap indices.
+		return $this->build_all_page_sitemap_indices();
+	}
+
+	/**
+	 * Build and store all page sitemaps.
+	 *
+	 * Side effect: Create/update sitemap rows sitemap-1, sitemap-2, etc.
+	 *
+	 * @access private
+	 * @since 4.6.0
+	 *
+	 * @return int The number of page sitemaps generated.
+	 */
+	private function build_all_page_sitemaps() {
```

Comment:
It is possible to consolidate the `build_all_` with a higher order function that takes a few extra arguments. I chose not to do that, or rather chose to postpone doing that, until it became more clear what the right abstraction was. (Because, and I'm new to the language so this may be inexperience talking, but that kind of abstraction seems to be expensive in PHP, and picking the wrong one too early looked more painful than having repetitive but straightforward code. We won't be adding new cases very often.)

For instance, in an earlier version, it also would have been possible to consolidate the `build_one_` functions in the same way. But I held off, and was glad later on when each one gathered little unique hairs (the filters) that would have been painful to add to a single higher order function.

**That said**, there are now enough examples that the right abstraction for `build_all_` may be more clear now.